### DISCUSSION TITLE
Alternative to @InjectMocks? (simple @Set?)

### DISCUSSION THREAD

**Discussion by nineninesevenfour** [mockito/mockito#3011] [Ideas]
Dear Mockitoans,

I won't bother you how I came to this topic. However I dug a little deeper into `@InjectMocks` and how it works and found some conceptual issues.

What additionally made me think is this comment on #1066

> We are not aiming to be a full injection framework

I strongly agree to this statement.

### The issues

I hope you find the following words not too offensive. This is not my intention. I rather want you to rethink the chosen path.

Here is the list of issues I currently see:
* It allows no precise selection of what is actually injected and where. It works rather like "take all the mocks around and toss it into an object with by some rules the user can hardly influence".
This is somehow contradictory to the mocking functionality itself, where the Mockito user has fine grained control over what happens and when.
* Injection is performed regardless of whether it is actually indicated through `@Inject` (CDI) or `@Autowire` (Spring) annotations or not.
  Although this might be seen as a feature, it actually depends on the use case, if this is desirable or not.
* It does not respect qualifiers at all.
* It does not (and also cannot and should not) mimic the semantics of the concrete injection framework used for productive code (e.g. scopes, lifecycle, bean discovery, etc.). However the name `@InjectMocks` somehow lures the user with unfulfillable expectations IMHO.
* when applied to generics it assumes that order and number of generic parameters are the same for the instance injected and the injection point, which is not necessarily the case.
(causing for example issues like #3006 and #3000).
* it cannot be applied recursively, meaning one cannot inject something that is itself annotated with `@InjectMocks` (only with additional tricks)
* only mocks can be injected
* it introduced quite a lot of additional code and coupling into Mockito that needs to be constantly respected and maintained

I have to admit, that `@InjectMocks` is suitable for simple cases, where it works well.

### The proposal

I propose to introduce a new (helper) annotation `@Set`, which allows to put (not necessarily mocked) values into private fields. It works very simple compared to `@InjectMocks`.

A short example:
```java
    // production code
    interface SomeBean { /*...*/ }
    
    class UseBean {
        @Inject
        private SomeBean someBean;
    }

    // test code
    @Mock(name = "mock")
    private SomeBean someBean;

    @Set(field = "someBean", source = "mock")
    private UseBean useBean = new UseBean();
```
The `@Set` annotation would then be processed during `MockitoAnnotations.openMocks(testInstance)` or the corresponding place when using `MockitoExtension`, `MockitoJunitRunner` or alike.

The parameters so far are:
* `field`: the name of the field in the target object
* `source`: the name of the source object (`@Mock` or `@Spy`) to  use. More on that right away.
It could be optional and if omitted the strategy is to use the field name for source lookup.

If someone wants to use a source, that is neither a spy nor a mock, I could imagine to additionally introduce an annotation `@Alias` to define such sources. For example:
```java
    @Alias(name = "notMocked")
    private SomeBean someBean = new SomeBeanImpl();
    @Set(field = "someBean", source = "notMocked")
    private UseBean useAnotherBean = new UseBean();
```
Or in the more shorter form:
```java
    @Alias
    private SomeBean someBean = new SomeBeanImpl();
    @Set(field = "someBean")
    private UseBean useAnotherBean = new UseBean();
```
### Postponed (manual) Processing
However there might come up the need to process `@Set` at a later point in time independently from `openMocks`. Especially when mocked objects are needed in the chain of objects to build up for testing.

Therefor I would suggest an additional parameter `automatic` (default value is `true`) and a corresponding method `MockitoAnnotations.doSetFields(testInstance)` to trigger processing of all `@Set` annotations that have set automatic=false.
For example:
```java
    // production code
    class ComplexBean {
        @Inject
        public ComplexBean(SomeBean someBean) {
            // ...
        }
    }

    class UseComplexBean {
        @Inject
        private ComplexBean complexBean;
    }

    // test code
    @Mock
    private SomeBean someBean;

    @Alias(name = "needsManualConstruction")
    private ComplexBean complexBean;

    @Set(field = "complexBean", source = "needsManualConstruction", automatic = false)
    private UseComplexBean useComplexBean = new UseComplexBean();

    private AutoCloseable closeMe, closeMeToo;

    @Before
    public void setup() {
        closeMe =  MockitoAnnotations.openMocks(this);
        Mockito.when(someBean.doesSomething()).thenReturn("did it");
        // can construct this only after "openMocks"
        complexBean = new ComplexBean(someBean);
        // trigger processing @Set(..., automatic = false)
        closeMeToo = MockitoAnnotations.doSetFields(this);
    }

    @After
    public void teardown() throws Exception {
        closeMe.close();
        closeMeToo.close();
    }

    @Test
    public void test() {
        // ...
    }
```
I am not sure about the names yet (please add your suggestions), but I think you've got the point.
### Primitive Types
When thinking it through, also setting primitive types might be appreciated like so:
```java
@Set(field = "theAnswer", value = "42")
```
There is still a lot of room for additional thoughts and ideas.
### The question
Strictly spoken  `@Set` has nothing to do with mocking. So the basic question is, if Mockito is willing to pack additional items into its code that simply supports writing testing code?

If that's the case, I could prepare a corresponding PR.

### About Constructor Injection
While elaborating I was tempted to suggest an additional...
```java
@Construct(sources = ..., types = ...)
```
... but I came to the conclusion, that it is better to leave construction for such cases up to the user similar to the example with `@Set(...automatic = false)` and `MockitoAnnotations.doSetFields(...)` above.

I am looking forward to your feedback! 😉

**Comment by nineninesevenfour** [comment]
TLDR; It's all about private fields and not to end up with something like `@InjectMocks(mode=WELD)` and `@InjectMocks(mode=SPRING)`. 😉

**Comment by ascopes** [comment]
I only glanced over this as I am on my phone, but for complex cases, would this be clearer and easier to read than explicitly initialising the constructor?

```java
@ExtendWith(MockitoExtension.class)
class TestFoo {
  Foo foo;
  @Mock Bar bar;
  @Mock Baz baz;
  @Mock Bork bork;

  @BeforeEach
  void setUp() {
    foo = new Foo(bar, baz, bork);
  }
}
```

Guess my main concern is whether there are cases that would make tests harder to maintain via more complicated wiring between components. The problem with annotations is you increasingly defer more and more sanity checking until runtime, so begin to miss out on existing checks that compilers and IDEs can already perform without needing specific support for Mockito static analysis.

Example: a test using the snippet above would fail compilation if I misspelled a parameter. If I were using strings, it would only detect the typo once JUnit runs.

Worth noting Spring provides a mechanism for mocking beans via injection in spring-test as well.

**Comment by nineninesevenfour** [comment]
@ascopes I agree that when one is the maintainer of a class (and the injection framework allows to do so), it is better to create the corresponding constructors. I also acknowledge your mentioned concerns.

I am talking about situations where you need to construct instances of classes as prerequisite for you test that either are not maintained by you or you have other strong reasons not to touch the existing code.

The problem I have with `@InjectMocks` is that Mockito decides for you which instances are appropriate to inject. For regular mocks you have fine-grained control on how your mocks are put together. With `@InjectMocks` you don't.
